controller/controller_test.go

```diff
@@ -224,6 +238,40 @@ func TestRunOnce(t *testing.T) {
 	assert.Equal(t, math.Float64bits(1), valueFromMetric(verifiedAAAARecords))
 }
 
+// TestRun tests that Run correctly starts and stops
+func TestRun(t *testing.T) {
+	source := getTestSource()
+	cfg := getTestConfig()
+	provider := getTestProvider()
+
+	r, err := registry.NewNoopRegistry(provider)
+	require.NoError(t, err)
+
+	// Run our controller once to trigger the validation.
+	ctrl := &Controller{
+		Source:             source,
+		Registry:           r,
+		Policy:             &plan.SyncPolicy{},
+		ManagedRecordTypes: cfg.ManagedDNSRecordTypes,
+	}
+	ctrl.nextRunAt = time.Now().Add(-time.Millisecond)
+	ctx, cancel := context.WithCancel(context.Background())
+	stopped := make(chan struct{})
+	go func() {
+		ctrl.Run(ctx)
+		close(stopped)
+	}()
+	time.Sleep(1500 * time.Millisecond)
```

Comment:
It's often a source of flakyness to introduce time.sleep in a test.
Do you think it's possible to use a readiness-like approach ?

Reply:
while you are right, it's not really possible to do so.
cancel() starts shutdown (triggers being non-ready) and we want to wait for an iteration of external-dns.
We could of course also create a hook inside external-dns control loop, but that's a bit too much for only testing IMO.